src/deluge/gui/views/automation_instrument_clip_view.cpp

```diff
@@ -0,0 +1,3138 @@
+
+			uint32_t squareStart = getPosFromSquare(xDisplay);
+
+			int32_t effectiveLength;
+
+			if (instrument->type == InstrumentType::KIT && !instrumentClipView.getAffectEntire()) {
+				ModelStackWithNoteRow* modelStackWithNoteRow = clip->getNoteRowForSelectedDrum(modelStack);
+
+				effectiveLength = modelStackWithNoteRow->getLoopLength();
+			}
+			else {
+				effectiveLength = clip->loopLength;
+			}
+
+			if (squareStart < effectiveLength) {
+
+				automationInstrumentClipView.interpolationBefore = false;
+				automationInstrumentClipView.interpolationAfter = false;
+
+				int32_t newKnobPos = calculateKnobPosForSinglePadPress(yDisplay);
+				setParameterAutomationValue(modelStackWithParam, newKnobPos, squareStart, xDisplay, effectiveLength);
+			}
+		}
+	}
+
+	uiNeedsRendering(this);
+}
+
+//calculates what the new parameter value is when you press a single pad
+int32_t AutomationInstrumentClipView::calculateKnobPosForSinglePadPress(int32_t yDisplay) {
+
+	int32_t newKnobPos = 0;
+
+	//if you press bottom pad, value is 0, for all other pads except for the top pad, value = row Y * 18
+	if (yDisplay >= 0 && yDisplay < 7) {
+		newKnobPos = yDisplay * kParamValueIncrementForAutomationSinglePadPress;
+	}
+	//if you are pressing the top pad, set the value to max (127)
+	else {
+		newKnobPos = kMaxKnobPos;
+	}
+
+	//in the deluge knob positions are stored in the range of -64 to + 64, so need to adjust newKnobPos set above.
+	newKnobPos = newKnobPos - kKnobPosOffset;
+
+	return newKnobPos;
+}
+
+//takes care of setting the automation values for the two pads pressed and the pads in between
+void AutomationInstrumentClipView::handleMultiPadPress(ModelStackWithTimelineCounter* modelStack, InstrumentClip* clip,
+                                                       int32_t firstPadX, int32_t firstPadY, int32_t secondPadX,
+                                                       int32_t secondPadY) {
+
+	Instrument* instrument = (Instrument*)clip->output;
+
+	if (modelStack) {
+
+		//calculate value corresponding to the two pads that were pressed in a multi pad press action
+		int32_t firstPadValue = calculateKnobPosForSinglePadPress(firstPadY) + kKnobPosOffset;
+		int32_t secondPadValue = calculateKnobPosForSinglePadPress(secondPadY) + kKnobPosOffset;
+
+		ModelStackWithAutoParam* modelStackWithParam =
+		    getModelStackWithParam(modelStack, clip, clip->lastSelectedParamID, clip->lastSelectedParamType);
+
+		if (modelStackWithParam && modelStackWithParam->autoParam) {
+
+			//if you want to enter long presses backwards, swap the first pad pressed with the second pad pressed
+			if (secondPadX < firstPadX) {
+
+				int32_t temp;
+				temp = firstPadX;
+				firstPadX = secondPadX;
+				secondPadX = temp;
+
+				temp = firstPadValue;
+				firstPadValue = secondPadValue;
+				secondPadValue = temp;
+			}
+
+			//loop from the firstPad to the secondPad, setting the values in between
+			for (int32_t x = firstPadX; x <= secondPadX; x++) {
```

Comment:
Do we actually need to generate all the intermediate nodes, or can those be implicitly generated (at least when interpolation is on)

Reply:
The interpolation setting I'm taking advantage of here in the deluge interpolates between nodes (e.g. if you set a value at the current node with the interpolation before setting on, it will interpolate the value from the previous node to this node.

so in theory, if I had only the first node and the second node created, it will probably interpolate the value from the first node to the second node.

but as soon as you start creating nodes in between those those first and last nodes, it's only going to interpolate between the first node to second node, second node to third node, third to fourth, and so on.

So the only way to be truly certain that it sets the correctly values for all the nodes is to "help" the interpolation out by giving it the values to interpolate between.

So if you think of the grid:

X0 X1 X2 X3 X4 X5 X6 X7 X8 X9 X10 X11 X12 X13 X14 X15

I start by activating interpolation

then if the node is X0, i disable interpolation before X0 (because I don't want to interpolate from X15 to X0)
if the node is X15, i disable interpolation after X15 (because again I don't want it to interpolate to X0)

for the nodes in between, interpolation is on

so if I supply X1, it will interpolate from the X0 value to the X1 value, and it's doing this at the very lowest level. 

E.g. if you are at the highest zoom level, you're really only supply 2 data points. But underneath it all, there wayyyyy more data points (sub-steps if you want to call them) and the deluge gradually bumps the values of those sub steps so that it's not a sudden jump in value from X0 to X1.

Technically what the deluge is doing is the same thing that I could do if I amended by code to work at the "node" level.

you can see in my code a call to "getPosFromSquare(x)" which is how the deluge converts the pads on the grid to node position.

In my code I'm only ever supplying the deluge with 16 nodes max (width of the grid). Underneath tho, there are way more nodes. At the top of the zoom level, 1 square = width of 24. At the lowest zoom level, 1 square = width of 3.

--

All this to say...I do need to loop through the firstPadX to secondPadX to make sure the values get set properly.

Reply:
I see, it's probably safest to keep doing things this way for now. I'll have to think about whether we can represent things more efficiently if/when I get around to refactoring how AutoParams store their automation.